Tidy up ResolverOpts usage, default trait impls

cpu opened the pull request:
The configuration situation for name servers & name server pools was complicated by the pervasive use of the `ResoverOpts` struct. This struct has a lot of configuration options, but very few are meaningfully used by these lower layers, resulting in some confusing behaviour (e.g. unit tests setting fields that have no effect on the actual implementation). For instance, `name_server_pool_tests.rs` tests were setting `try_tcp_on_error` at a layer where the value was never consulted, and the `recursor_dns_handle.rs` `recursor_opts()` func was setting/documenting fields like `validate` that had no effect.

I suspect there's further room for improvement; in particular, both before & after this branch we're using hard-coded values for `NameServerOptions` configuration when building the root server pool, and when creating a name server pool for a zone. For now since this is a large diff already I've called this out with a couple of TODO comments while preserving the values used previously.

Along the way I also took this as a chance to tidy up a handful of manual `Default` impls on enums that could be replaced with derived implementations by marking the correct enum variant as the `#[default]`.

cpu committed:
config: introduce ConnectionOpts type

Previously the `ConnectionProvider` trait was taking a `&ResolverOpts`
arg in `new_connection()`, but in practice only ~3 fields are used.

This commit introduces a new type in `config` for the connection
specific configuration connection providers require, and embeds it in
the larger `ResolverOpts` struct as a flattened field.

This is groundwork for simplifying the recursor code that was having to
make the wider `ResolverOpts` struct just for the `ConnectionProvider`.

cpu committed:
recursor: extract case randomization to constructor arg

This will make it easier to simplify the options arguments in
a subsequent commit. The case randomization bool is only used shallowly,
and so can be set separately from the configuration that's passed down
to deeper layers as part of an opts struct.

cpu committed:
replace manual Default impl for enums

Use the `#[default]` syntax on a variant and a derived `Default`
instead.

djc reviewed (commented):
My brain is a little fried but here's some early feedback.

I'm not fully convinced the introduction of `ConnectionOptions` and `NameServerOptions` is the right way forward. This seems to fragment/duplicate the configuration within the resolver crate. I would find it more convincing if we could remove the fields from `ResolverOpts` and/or use the smaller types inside `ResolverOpts` as well.

(One approach I've been thinking about is that quite a few things in `ResolverOpts` are actually used at the proto level and so it might make sense to extract a proto-level config type out of the `ResolverOpts`. Haven't looked into it much, though...)

Also, might still be useful to isolate the straightforward parts into a separate PR so we can get them merged?

djc commented on crates/resolver/src/name_server/connection_provider.rs:
```diff
@@ -270,6 +271,49 @@ impl TlsConfig {
 
+/// Connection options for creating a new connection with a connection provider.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ConnectionOptions {
```
It's a bit of a kludge for sure. Happy to consider something better if you can come up with something, but so far I haven't managed to despite spending a decent amount of time on it.

cpu committed:
resolver: split out NameServerOptions from ResolverOpts

Lift `ConnectionOptions` into `NameServerOptions`, flatten both into
`ResolverOpts` for serde-purposes.

This will make subsequent commits to pass through more narrow parts of
the config to deeper layers possible to achieve.

cpu committed:
use NameServerOptions for resolver name servers/pools

When building name servers and pools of name servers it's confusing to
work with the complex `ResolverOpts` type, especially in the context of
the recursive resolver where many of the options are not used. Instead,
introduce a thinner `NameServerOptions` type that the relevant parts of
the code can use.

cpu commented on crates/resolver/src/name_server/connection_provider.rs:
```diff
@@ -270,6 +271,49 @@ impl TlsConfig {
+            avoid_local_udp_ports: opts.avoid_local_udp_ports.clone(),
+        }
+    }
+}
```
Done.

cpu commented on crates/resolver/src/name_server/name_server.rs:
```diff
@@ -272,6 +272,46 @@ impl<P: ConnectionProvider> NameServerState<P> {
 
+/// Options specific to interacting with name servers.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NameServerOptions {
```
Fixed.

cpu commented on crates/resolver/src/config.rs:
```diff
@@ -443,7 +443,7 @@ impl ProtocolConfig {
 #[cfg_attr(
     feature = "serde",
     derive(Serialize, Deserialize),
-    serde(default, deny_unknown_fields)
```
I had to remove `deny_unknown_fields` here, or `config_tests::test_read_config` finds `timeout` as an unknown field, even though it comes from the flattened `ConnectionOptions` (or `NameServerOptions.ConnectionOptions.timeout` in subsequent commits).

Removing `deny_unknown_fields` also required a `#[ignore]` on the `test_reject_unknown_fields()` test. If we decide to go without `deny_unknown_fields`, I'll delete it. Otherwise I'll have to un-flatten a layer in the TOML I think.

djc replied:
I don't think we want to get rid of `deny_unknown_fields`, it seems important to me.

djc commented on crates/resolver/src/config.rs:
```diff
@@ -584,6 +570,39 @@ fn default_recursion_desired() -> bool {
 }
 
+/// Options specific to interacting with name servers.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize), serde(default))]
```
This should be `non_exhaustive`.

djc commented on tests/integration-tests/tests/integration/name_server_pool_tests.rs:
```diff
@@ -665,7 +685,11 @@ fn test_user_provided_server_order() {
         vec![preferred_nameserver, secondary_nameserver],
         None,
-        options,
+        NameServerOptions {
```
Assuming this syntax still works when `NameServerOptions` becomes non-exhaustive, suggest an intermediate commit that transitions to the partial update syntax in a separate commit from the one changing from `ResolverOpts` to `NameServerOptions`.

djc reviewed (commented):
This still seems pretty messy in its current state. Which parts are actually important for your larger goals?

It does seem the `NameServerOptions` extraction (as a type that encompasses the parts of `ResolverOpts` that are only relevant to `NameServerPool` and `NameServer`) probably makes sense. I think it's probably also okay for that to be an explicitly nested type inside `ResolverOpts` at the config level.
